### DISCUSSION TITLE
Wait for MongoDB collections creation in Swarm environment

### DISCUSSION THREAD

**Discussion by Stramba** [BretFisher/ama#187] [Q&A]
I have a NodeJS server which have to connect to a MongoDB instance, both running as docker containers in a Swarm environment. 

The database contains collections created and filled at startup using the default procedure explained in its image repo. The server connects to database at startup to read configuration parameters.

I would like to ensure that NodeJS service will start when database completed its initialization (and not as soon as it is "up & running"), but I do not understand how should I do this in Swarm. The only options I thought are:
 - split deployments and ensure to wait for database initialization before starting the server
 - deploy all together, but forcing the server to quit and restart as long as queries to database throw exceptions or return empty collections

Is there any other approach that I should use to handle this kind of "synchronization"?

**Comment by BretFisher** [comment]
In any distributed system, your applications should be able to handle a remote dependency going down, because that's a common occurrence. Database drivers and libraries have retry built-in usually, and your app can either try to handle connection errors itself (with retry) or just exit with an error code and let the underlying infrastructure handle it.

Swarm and Kubernetes do this by seeing an exit (with error code) and tries to restart the container or replace it.

**Comment by Stramba** [comment]
My question (probably bad written) was not related to database unavailability, but _data unavailability_: how should I handle that MongoDB service is up but the script filling it has not yes finished? 

So my problem is not related to failed connections, because database could be up and the app can connect to it; the problem is that the app could start, query the db, receive empty data and still live. That's because queried data may be just decorative information (description of objects or associations) and the app logic could be fine also without them.

I could handle it at app code level (if the query returns empty data just exit and retry) but this converts data from optional to mandatory.  I would like to know if there are "OPS" best practices to handle the initial filling of a deployed database.

**Comment by BretFisher** [comment]
OK, no this is not an ops problem really. Cloud Native apps should be able to handle common data states in a predictable way.  Those ways include things like:

* db connection unavailable
* db login failed
* db connection and login work, but database or table missing (error on query)
* returned data was not what was expected (either bad data or mismatch of schema versions)

If your app currently `exit(1)` when it doesn't receive the data it expects, and you're happy with that, then fine. If you'd like a more user (and dev) friendly way, then catch that exception and provide more detail to the dev/ops/user/etc. Generic errors are usually not obvious to people who aren't experts in that framework/app/service, so more user-friendly errors are always welcome at every level.

For example, there are some well-written apps that will say something like "db login successful but database is empty" or even guide them to the next step for fixing. Putting this level of effort into your server application design is something most don't do, which is unfortunate because dealing with "unexpected app states" is a sign of a mature codebase. It shows that it has gone through development efforts for planning all of the potential states.
